File: backend/database/models/api_key.go
package models

import (
	"time"

	"github.com/guregu/null"
)

type ApiKeyAccess string

const (
	ApiKeyAccessNone  ApiKeyAccess = "none"
	ApiKeyAccessRead  ApiKeyAccess = "read"
	ApiKeyAccessWrite ApiKeyAccess = "write"
	ApiKeyAccessFull  ApiKeyAccess = "full"
)

type ApiKey struct {
	ID          string       `json:"id" db:"id"`
	CreatedBy   string       `json:"created_by" db:"created_by"`
	AppID       string       `json:"app_id" db:"app_id"`
	Name        string       `json:"name" db:"name"`
	Description null.String  `json:"description" db:"description"`
	Access      ApiKeyAccess `json:"access" db:"access"`
	ExpiredAt   null.Int     `json:"expired_at" db:"expired_at"`
	LastUsed    null.Time    `json:"last_used" db:"last_used"`
	Metadata    interface{}  `json:"metadata" db:"metadata"`
	CreatedAt   time.Time    `json:"created_at,omitempty" db:"created_at,omitempty"`
	UpdatedAt   time.Time    `json:"updated_at,omitempty" db:"updated_at,omitempty"`
	DeletedAt   null.Time    `json:"deleted_at,omitempty" db:"deleted_at"`
}

File: frontend/app/models/api_key.ts
import { z } from "zod";
import { ServerResponseSchema } from "./default";

export const ApiKeySchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable().transform((value) => value ?? ""),
  access: z.string(),
  expired_at: z.number().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  last_used: z.string().nullable().optional(),
  key_prefix: z.string().optional(),
  last_four: z.string().optional(),
  secret: z.string().optional(),
});

export type ApiKey = z.infer<typeof ApiKeySchema>;

export const ApiKeysSchema = z.array(ApiKeySchema);

export type ApiKeys = z.infer<typeof ApiKeysSchema>;

export const GetApiKeySchema = ServerResponseSchema.extend({
  data: ApiKeySchema,
});

export const GetApiKeysSchema = ServerResponseSchema.extend({
  data: ApiKeysSchema,
});

export const CreateApiKeyFormSchema = z.object({
  name: z.string().min(3),
  description: z.string().optional(),
  access: z.enum(["read", "write", "full"]).default("read"),
  intent: z.enum(["create", "update", "delete"]).default("create"),
  id: z.string().optional(),
});

export type CreateApiKeyForm = z.infer<typeof CreateApiKeyFormSchema>;
